db/migrate/20240903024211_add_trigger_to_users_to_sync_user_passwords.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+class AddTriggerToUsersToSyncUserPasswords < ActiveRecord::Migration[7.1]
+  def up
+    # necessary for postgres < v14 which does not have CREATE OR REPLACE TRIGGER
+    execute <<~SQL.squish
+      DROP TRIGGER IF EXISTS
+      users_password_sync ON users;
+    SQL
+
+    execute <<~SQL.squish
+    CREATE OR REPLACE FUNCTION mirror_user_password_data() RETURNS TRIGGER AS $$
+    BEGIN
+      INSERT INTO user_passwords (user_id, password_hash, password_salt, password_algorithm, password_expired_at, created_at, updated_at)
+      VALUES (NEW.id, NEW.password_hash, NEW.salt, NEW.password_algorithm, NULL, now(), now())
+      ON CONFLICT(user_id)
+      DO UPDATE SET
+        password_hash = EXCLUDED.password_hash,
+        password_salt = EXCLUDED.password_salt,
+        password_algorithm = EXCLUDED.password_algorithm,
+        password_expired_at = EXCLUDED.password_expired_at,
+        updated_at = now()
+        WHERE user_passwords.password_hash <> EXCLUDED.password_hash;
+      RETURN NEW;
+    END;
+    $$ LANGUAGE plpgsql;
+    SQL
+
+    execute <<~SQL.squish
+      CREATE TRIGGER users_password_sync
+      AFTER INSERT OR UPDATE OF password_hash ON users
+      FOR EACH ROW
+      WHEN (NEW.password_hash IS NOT NULL)
```

Comment:
What about if someone is deliberately nulling out a user's password on the console? I guess we should mirror that as a deletion of the relevant user_passwords row?

Reply:
Ok yeah, that makes sense. Agree it's not worth the effort if it's only for the console, and it's only for the duration of the deploy :+1: